### ISSUE TITLE
Stack overflow when redefining a variable

### ISSUE THREAD

**Issue by csyonghe** [shader-slang/slang#375]
I am not sure whether this has already been reported as an issue, but the following triggers stack overflow:
```
StructType values;
void test()
{
    InnerStructType values = values.inner;
}
```

**Comment by tfoleyNV** [comment]
My guess is that the reference to `values` on the right-hand side is trying to look up the definition of `values` inside the function itself, and triggering all our logic that tries to check declarations on-demand.

There is a question here of whether this should be illegal, because it is circular, or whether it should be allowed to work, so that a local variable should check its initializer in the scope from *before* it gets declared. The latter is probably what we need to do to fix some other bugs we have, but it will be annoying to implement.

**Comment by tfoleyNV** [comment]
Should be fixed by #385.
